File: polyglot/typescript/binary_parser.ts
import * as fs from 'fs';
import * as path from 'path';

// ============================================================================
// TYPES & INTERFACES
// ============================================================================

interface PackerInfo {
  name: string;
  magic: number[];
  version?: string;
  confidence: number; // 0-100
}

interface EntropyResult {
  totalBits: number;
  entropyPerByte: number;
  isHighEntropy: boolean;
}

interface PEHeaderInfo {
  machine: string;
  numberOfSections: number;
  timestamp: Date | null;
  subsystem: string;
  characteristics: string[];
}

interface AnalysisResult {
  filename: string;
  sizeBytes: number;
  magicNumbers: Map<number, PackerInfo>;
  entropy: EntropyResult;
  peHeader: PEHeaderInfo | null;
  yaraRules: string[];
  sarifResults: SarifResultItem[];
}

interface SarifResultItem {
  id: string;
  ruleId: string;
  level: 'note' | 'warning' | 'error';
  message: string;
  locations?: Array<{
    physicalLocation: {
      artifactLocation: { uri: string };
    };
  }>;
}

// ============================================================================
// CONSTANTS & MAGIC NUMBERS
// ============================================================================

const PE_MAGIC = 0x4D5A; // "MZ"
const UPX_MAGIC = [0x55, 0x54]; // "UT"
const ASPACK_MAGIC = [0x21, 0x41, 0x53, 0x50]; // "!ASP"
const THEMIDA_MAGIC = [0x74, 0x68, 0x6D, 0x64]; // "thmd"
const VMProtect_MAGIC = [0x56, 0x4D, 0x50, 0x31]; // "VMP1"
const MPRESS_MAGIC = [0x4D, 0x50, 0x52, 0x45]; // "MPRE"

const PE_HEADER_OFFSET = 64;
const PE_MACHINE_OFFSET = 180;
const PE_SUBSYSTEM_OFFSET = 182;

// ============================================================================
// UTILITY FUNCTIONS
// ============================================================================

function readUInt16LE(buffer: Buffer, offset: number): number {
  return (buffer[offset] | buffer[offset + 1] << 8);
}

function readUInt32LE(buffer: Buffer, offset: number): number {
  let result = 0;
  for (let i = 0; i < 4; i++) {
    result |= buffer[offset + i] << (i * 8);
  }
  return result;
}

function readStringLE(buffer: Buffer, offset: number, length: number): string {
  let str = '';
  for (let i = 0; i < length && i < buffer.length - offset; i++) {
    const charCode = buffer[offset + i];
    if (charCode === 0) break;
    str += String.fromCharCode(charCode);
  }
  return str.trim();
}

function calculateEntropy(buffer: Buffer): EntropyResult {
  // Calculate Shannon entropy per byte
  const frequency: number[] = new Array(256).fill(0);
  
  for (let i = 0; i < buffer.length; i++) {
    frequency[buffer[i]]++;
  }

  let totalBits = 0;
  let maxFrequency = 0;

  for (const count of frequency) {
    if (count > 0) {
      const probability = count / buffer.length;
      totalBits += -probability * Math.log2(probability);
      maxFrequency = Math.max(maxFrequency, count);
    }
  }

  // Normalize to bits per byte (max is ~8 for uniform distribution)
  const entropyPerByte = totalBits / 8;

  return {
    totalBits: Math.round(totalBits * 100) / 100,
    entropyPerByte: Math.round(entropyPerByte * 100) / 100,
    isHighEntropy: entropyPerByte > 6.5, // Threshold for packed/encrypted content
  };
}

function detectPEHeader(buffer: Buffer): PEHeaderInfo | null {
  if (buffer.length < PE_HEADER_OFFSET + 2 || readUInt16LE(buffer, 0) !== PE_MAGIC) {
    return null;
  }

  const machine = readUInt16LE(buffer, PE_MACHINE_OFFSET);
  let subsystem: string;
  
  switch (machine) {
    case 0x14c: subsystem = 'Intel x86'; break;
    case 0x8664: subsystem = 'AMD64 / x64'; break;
    case 0xaa: subsystem = 'ARM'; break;
    default: subsystem = `Unknown (${machine.toString(16)})`;
  }

  const timestamp = readUInt32LE(buffer, PE_HEADER_OFFSET + 4);
  const timestampDate = timestamp > 0 ? new Date(timestamp * 1000) : null;

  // Parse characteristics flags
  const characteristics: string[] = [];
  if (readUInt16LE(buffer, PE_HEADER_OFFSET + 2) & 0x0001) {
    characteristics.push('Relocatable');
  }
  if (readUInt16LE(buffer, PE_HEADER_OFFSET + 2) & 0x0002) {
    characteristics.push('ExecutableImage');
  }
  if (readUInt16LE(buffer, PE_HEADER_OFFSET + 2) & 0x0004) {
    characteristics.push('LineNumbers');
  }
  if (readUInt16LE(buffer, PE_HEADER_OFFSET + 2) & 0x0008) {
    characteristics.push('WorkingSet');
  }
  if (readUInt16LE(buffer, PE_HEADER_OFFSET + 2) & 0x0010) {
    characteristics.push('ReadOnlyData');
  }
  if (readUInt16LE(buffer, PE_HEADER_OFFSET + 2) & 0x0020) {
    characteristics.push('SharedLibrary');
  }
  if (readUInt16LE(buffer, PE_HEADER_OFFSET + 2) & 0x0040) {
    characteristics.push('ExecutableCode');
  }

  return {
    machine: subsystem,
    numberOfSections: readUInt16LE(buffer, PE_HEADER_OFFSET),
    timestamp: timestampDate,
    subsystem: 'PE32', // Simplified for this demo
    characteristics,
  };
}

function detectPacker(buffer: Buffer): PackerInfo[] {
  const results: PackerInfo[] = [];

  // Check UPX (UT magic at offset 0x14)
  if (buffer.length >= 0x18 && 
      buffer[0x14] === UPX_MAGIC[0] && 
      buffer[0x15] === UPX_MAGIC[1]) {
    const version = readStringLE(buffer, 0x16, 2);
    results.push({
      name: 'UPX',
      magic: [UPX_MAGIC[0], UPX_MAGIC[1]],
      version: `v${version || 'unknown'}`,
      confidence: 95,
    });
  }

  // Check ASPack
  if (buffer.length >= 4 && 
      buffer[0] === ASPACK_MAGIC[0] &&
      buffer[1] === ASPACK_MAGIC[1] &&
      buffer[2] === ASPACK_MAGIC[2] &&
      buffer[3] === ASPACK_MAGIC[3]) {
    results.push({
      name: 'ASPack',
      magic: [APSPACK_MAGIC[0], ASPACK_MAGIC[1], ASPACK_MAGIC[2], ASPACK_MAGIC[3]],
      confidence: 98,
    });
  }

  // Check Themida
  if (buffer.length >= 4 && 
      buffer[0] === THEMIDA_MAGIC[0] &&
      buffer[1] === THEMIDA_MAGIC[1] &&
      buffer[2] === THEMIDA_MAGIC[2] &&
      buffer[3] === THEMIDA_MAGIC[3]) {
    results.push({
      name: 'Themida',
      magic: [THEMIDA_MAGIC[0], THEMIDA_MAGIC[1], THEMIDA_MAGIC[2], THEMIDA_MAGIC[3]],
      confidence: 96,
    });
  }

  // Check VMProtect
  if (buffer.length >= 4 && 
      buffer[0] === VMProtect_MAGIC[0] &&
      buffer[1] === VMProtect_MAGIC[1] &&
      buffer[2] === VMProtect_MAGIC[2] &&
      buffer[3] === VMProtect_MAGIC[3]) {
    results.push({
      name: 'VMProtect',
      magic: [VMProtect_MAGIC[0], VMProtect_MAGIC[1], VMProtect_MAGIC[2], VMProtect_MAGIC[3]],
      confidence: 97,
    });
  }

  // Check MPRESS
  if (buffer.length >= 4 && 
      buffer[0] === MPRESS_MAGIC[0] &&
      buffer[1] === MPRESS_MAGIC[1] &&
      buffer[2] === MPRESS_MAGIC[2] &&
      buffer[3] === MPRESS_MAGIC[3]) {
    results.push({
      name: 'MPRESS',
      magic: [MPRESS_MAGIC[0], MPRESS_MAGIC[1], MPRESS_MAGIC[2], MPRESS_MAGIC[3]],
      confidence: 94,
    });
  }

  return results;
}

function generateYaraRules(packers: PackerInfo[], entropy: EntropyResult): string[] {
  const rules: string[] = [];

  // UPX rule
  if (packers.some(p => p.name === 'UPX')) {
    rules.push(`
rule UPX_Detector {
  meta:
    author = "PackPeek"
    description = "Detects UPX packing"
    tags = "packing,upx"

  strings:
    $upx_magic = "UT" at offset 0x14 with length 2
    $upx_version = { "UPX!" "UPX0" } of ascii

  condition:
    uint16(0x14) == 0x5554 and 
    any of ($upx_magic, $upx_version)
}
`);
  }

  // ASPack rule
  if (packers.some(p => p.name === 'ASPack')) {
    rules.push(`
rule ASPack_Detector {
  meta:
    author = "PackPeek"
    description = "Detects ASPack packing"
    tags = "packing,aspack"

  strings:
    $aspack_magic = "!ASP" at offset 0x00 with length 4

  condition:
    uint8(0x00) == 0x21 and 
    uint8(0x01) == 0x41 and 
    uint8(0x02) == 0x53 and 
    uint8(0x03) == 0x50
}
`);
  }

  // High entropy rule (potential encryption/VMProtect)
  if (entropy.isHighEntropy) {
    rules.push(`
rule Potential_Virtualization {
  meta:
    author = "PackPeek"
    description = "Detects potentially virtualized or encrypted binary"
    tags = "entropy,virtualization,encryption"

  condition:
    uint16(0x00) == 0x4D5A and 
    float(entropy_per_byte) > 6.5
}
`);
  }

  return rules;
}

function generateSarifResults(filename: string, packers: PackerInfo[], entropy: EntropyResult): SarifResultItem[] {
  const results: SarifResultItem[] = [];

  // Add packer findings as notes
  for (const packer of packers) {
    results.push({
      id: `packer_${packer.name.toLowerCase()}`,
      ruleId: `PackPeek.${packer.name}_DETECTED`,
      level: 'note',
      message: `${packer.name} packing detected with ${Math.round(packer.confidence)}% confidence.`,
      locations: [{
        physicalLocation: {
          artifactLocation: { uri: filename },
        },
      }],
    });
  }

  // Add entropy finding as warning if high
  if (entropy.isHighEntropy) {
    results.push({
      id: 'high_entropy',
      ruleId: 'PackPeek.HIGH_ENTROPY',
      level: 'warning',
      message: `Binary shows high entropy (${entropy.entropyPerByte.toFixed(2)} bits/byte), suggesting encryption or virtualization protection.`,
      locations: [{
        physicalLocation: {
          artifactLocation: { uri: filename },
        },
      }],
    });
  }

  // Add PE header info as note if valid
  const peHeader = detectPEHeader(Buffer.from(filename, 'binary'));
  if (peHeader) {
    results.push({
      id: 'pe_header',
      ruleId: 'PackPeek.PE_HEADER_PARSED',
      level: 'note',
      message: `Valid PE header found. Machine: ${peHeader.machine}, Sections: ${peHeader.numberOfSections}.`,
      locations: [{
        physicalLocation: {
          artifactLocation: { uri: filename },
        },
      }],
    });
  }

  return results;
}

// ============================================================================
// MAIN PARSER CLASS
// ============================================================================

export class BinaryParser {
  private buffer: Buffer;
  private result: AnalysisResult | null = null;

  constructor(buffer: Buffer) {
    this.buffer = buffer;
  }

  analyze(): AnalysisResult {
    if (this.result) return this.result;

    const filename = 'input.bin'; // Default name when no file path provided
    const sizeBytes = this.buffer.length;

    // Detect packers
    const packers = detectPacker(this.buffer);

    // Calculate entropy
    const entropy = calculateEntropy(this.buffer);

    // Parse PE header if present
    let peHeader: PEHeaderInfo | null = null;
    if (readUInt16LE(this.buffer, 0) === PE_MAGIC) {
      peHeader = detectPEHeader(this.buffer);
    }

    // Generate YARA rules
    const yaraRules = generateYaraRules(packers, entropy);

    // Generate SARIF results
    const sarifResults = generateSarifResults(filename, packers, entropy);

    this.result = {
      filename: filename,
      sizeBytes: sizeBytes,
      magicNumbers: new Map(
        packers.map((p) => [p.magic[0], p])
      ),
      entropy: entropy,
      peHeader: peHeader,
      yaraRules: yaraRules,
      sarifResults: sarifResults,
    };

    return this.result;
  }

  getAnalysis(): AnalysisResult {
    if (!this.result) {
      throw new Error('Call analyze() first');
    }
    return this.result;
  }

  reset(): void {
    this.buffer = Buffer.from(this.buffer);
    this.result = null;
  }
}

// ============================================================================
// EXPORTED FUNCTIONS FOR EASE OF USE
// ============================================================================

export function analyzeBinaryFile(filePath: string): AnalysisResult {
  const buffer = fs.readFileSync(filePath);
  return new BinaryParser(buffer).analyze();
}

export function analyzeBuffer(buffer: Buffer): AnalysisResult {
  return new BinaryParser(buffer).analyze();
}

// ============================================================================
// DEMO / RUNNABLE ENTRY POINT
// ============================================================================

function main(): void {
  // Demo with a sample PE file (minimal valid MZ header)
  const demoBuffer = Buffer.from([
    0x4D, 0x5A, 0x90, 0x00, 0x03, 0x00, 0x00, 0x00, // MZ header
    0x04, 0x00, 0x0C, 0x00, 0x14, 0x00, 0x00, 0x00,
    0x02, 0x00, 0x00, 0x00, 0x40, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    // ... rest of minimal PE header
    0x

File: polyglot/ruby/binary_parser.rb
require 'json'
require 'digest/sha1'
require 'zlib'

module Packpeek
  VERSION = "0.1.0"

  # Magic bytes for known packers
  PACKER_SIGNATURES = {
    :upx => "\x4D\x50",           # UPX!
    :aspack => "\x53\x48\x49\x46\x4C\x4F\x57", # "ASPack"
    :themida => ["\x54\x68\x65\x6D\x69\x64\x61", "\x54\x68\x6D\x69\x64\x61"], # "Themida"/"Thmida"
    :mpress => "\x4D\x50\x52\x45\x53\x53", # MPRESS
    :vmprotect => ["\x56\x4D\x50\x72\x6F\x74\x65\x63\x74", "\x56\x4D\x50\x72"]
  }.freeze

  class BinaryParser
    attr_reader :file_path, :data, :headers, :entropy

    def initialize(file_path = nil)
      @file_path = file_path
      @data = nil
      @headers = {}
      @entropy = 0.0
    end

    # Load binary data from file or stdin
    def load!
      return self.data if @data

      if @file_path
        File.open(@file_path, "rb") do |f|
          @data = f.read
        end
      else
        @data = STDIN.read
      end

      unless @data.is_a?(String) && !@data.empty?
        raise ArgumentError, "Empty or invalid binary data"
      end

      self
    end

    # Parse PE headers if present
    def parse_pe_headers!
      return self.headers if @headers.any? { |_, v| v[:found] }

      offset = 0
      while offset < @data.length - 64
        header = @data[offset, 64].bytes.to_a
        break unless header.all? { |b| b >= 0 && b <= 255 }

        # Check for PE signature "MZ" at current offset
        if header[0] == 80 && header[1] == 90 # 'M' = 76, 'Z' = 90 (ASCII)
          break
        end

        offset += 32
      end

      @headers[:pe_offset] = offset - 64 if offset > 0
      parse_pe_at(offset) unless offset == 0
    end

    # Parse PE header at given offset
    def parse_pe_at(pe_offset)
      return self.headers if pe_offset >= @data.length

      data = @data[pe_offset, 256]
      bytes = data.bytes.to_a

      # DOS header check
      dos_magic = [bytes[0], bytes[1]]
      
      unless dos_magic == [79, 80] # "MZ"
        @headers[:dos_found] = false
        return self.headers
      end

      # Extract PE header info from DOS stub
      pe_offset_val = (bytes[62] << 8) + bytes[63]
      
      if pe_offset_val > 0 && pe_offset_val < @data.length
        pe_data = @data[pe_offset_val, 256]
        pe_bytes = pe_data.bytes.to_a

        # PE signature "PE\0\0"
        unless [pe_bytes[0], pe_bytes[1]] == [80, 69]
          @headers[:dos_found] = false
          return self.headers
        end

        # Machine type (Intel x86 = 0x14c, x64 = 0x8664)
        machine_type = (pe_bytes[2] << 8) + pe_bytes[3]
        
        @headers[:dos_found] = true
        @headers[:machine_type] = {
          0x14c => "Intel x86",
          0x8664 => "Intel x64"
        }[machine_type] || machine_type.to_s

        # PE header fields
        @headers[:pe_offset] = pe_offset_val
        @headers[:timestamp] = (pe_bytes[18] << 24) | (pe_bytes[19] << 16) | 
                               (pe_bytes[20] << 8) | pe_bytes[21]
        
        # Optional header size
        opt_header_size = (pe_bytes[30] << 8) + pe_bytes[31]

        if opt_header_size > 0 && pe_offset_val + 56 < @data.length
          opt_data = @data[pe_offset_val + 56, opt_header_size]
          opt_bytes = opt_data.bytes.to_a

          # Optional header magic (PE32+ vs PE32)
          opt_magic = [opt_bytes[0], opt_bytes[1]]
          
          if opt_magic == [245, 22] # PE32+
            @headers[:pe_type] = "PE32+"
            subheader_offset = 96
          else
            @headers[:pe_type] = "PE32"
            subheader_offset = 80
          end

          if pe_offset_val + subheader_offset < @data.length
            sub_data = @data[pe_offset_val + subheader_offset, 128]
            sub_bytes = sub_data.bytes.to_a

            # Sub-header fields (ImageBase, SectionAlignment, etc.)
            @headers[:image_base] = (sub_bytes[4] << 32) | 
                                   (sub_bytes[5] << 24) | 
                                   (sub_bytes[6] << 16) | 
                                   (sub_bytes[7] << 8) | sub_bytes[8]
            
            @headers[:section_alignment] = (sub_bytes[10] << 8) + sub_bytes[11]
            @headers[:file_alignment] = (sub_bytes[12] << 8) + sub_bytes[13]

            # Number of sections
            num_sections = (sub_bytes[16] << 8) + sub_bytes[17]
            
            if num_sections > 0 && pe_offset_val + subheader_offset + 4 < @data.length
              section_data = @data[pe_offset_val + subheader_offset + 4, 
                                   num_sections * 40]
              
              sections = []
              i = 0
              while i < num_sections && i < section_data.length / 40
                sec_bytes = section_data[i*40, 40].bytes.to_a
                
                # Section name (8 bytes)
                name = [sec_bytes[0], sec_bytes[1]].pack("C2").upcase
                sections << {
                  :name => name,
                  :virtual_size => (sec_bytes[16] << 24) | 
                                  (sec_bytes[17] << 16) | 
                                  (sec_bytes[18] << 8) | sec_bytes[19],
                  :raw_size => (sec_bytes[20] << 24) | 
                             (sec_bytes[21] << 16) | 
                             (sec_bytes[22] << 8) | sec_bytes[23],
                  :virtual_address => (sec_bytes[24] << 24) | 
                                    (sec_bytes[25] << 16) | 
                                    (sec_bytes[26] << 8) | sec_bytes[27],
                  :raw_offset => (sec_bytes[32] << 24) | 
                               (sec_bytes[33] << 16) | 
                               (sec_bytes[34] << 8) | sec_bytes[35]
                }

                i += 1
              end

              @headers[:sections] = sections
            end
          end
        end
      end

      self.headers
    end

    # Detect packer signatures
    def detect_packer!
      return self.headers if @headers.key?(:packers)

      data = @data.dup
      
      # UPX detection (check multiple offsets)
      upx_found = []
      [0, 64].each do |offset|
        next unless offset + 2 < data.length
        magic = [data[offset], data[offset+1]]
        if magic == [76, 80] # "UP"
          upx_found << { :offset => offset, :type => :upx }
        end
      end

      # ASPack detection
      aspack_found = []
      [0, 64].each do |offset|
        next unless offset + 7 < data.length
        magic = [data[offset], data[offset+1], data[offset+2], 
                data[offset+3], data[offset+4], data[offset+5], 
                data[offset+6]]
        if magic == [83, 72, 73, 70, 76, 79, 87] # "ASPack"
          aspack_found << { :offset => offset, :type => :aspack }
        end
      end

      # Themida detection
      themida_found = []
      [0, 64].each do |offset|
        next unless offset + 7 < data.length
        magic1 = [data[offset], data[offset+1], data[offset+2], 
                 data[offset+3], data[offset+4], data[offset+5], 
                 data[offset+6]]
        magic2 = [data[offset], data[offset+1], data[offset+2], 
                 data[offset+3], data[offset+4], data[offset+5], 
                 data[offset+6] - 1] # Check for "Thmida" variant
        
        if magic1 == [84, 72, 101, 109, 105, 100, 97] ||
           magic2 == [84, 72, 101, 109, 105, 100, 97]
          themida_found << { :offset => offset, :type => :themida }
        end
      end

      # MPRESS detection
      mpress_found = []
      [0, 64].each do |offset|
        next unless offset + 6 < data.length
        magic = [data[offset], data[offset+1], data[offset+2], 
                data[offset+3], data[offset+4], data[offset+5]]
        if magic == [76, 80, 82, 69, 83, 83] # "MPRESS"
          mpress_found << { :offset => offset, :type => :mpress }
        end
      end

      # VMProtect detection
      vmprotect_found = []
      [0, 64].each do |offset|
        next unless offset + 8 < data.length
        magic1 = [data[offset], data[offset+1], data[offset+2], 
                 data[offset+3], data[offset+4], data[offset+5], 
                 data[offset+6], data[offset+7]]
        magic2 = [data[offset], data[offset+1], data[offset+2], 
                 data[offset+3], data[offset+4], data[offset+5], 82, 80] # "VMPro"
        
        if magic1 == [86, 77, 80, 114, 111, 116, 101, 99] ||
           magic2 == [86, 77, 80, 114, 111, 116, 82, 116]
          vmprotect_found << { :offset => offset, :type => :vmprotect }
        end
      end

      @headers[:packers] = {
        upx: upx_found.any?,
        aspack: aspack_found.any?,
        themida: themida_found.any?,
        mpress: mpress_found.any?,
        vmprotect: vmprotect_found.any?
      }

      self.headers
    end

    # Calculate Shannon entropy for data
    def calculate_entropy!
      return @entropy if @entropy > 0
      
      data = @data.dup
      length = data.length
      
      unless length > 0
        @entropy = 0.0
        return self.entropy
      end

      # Count byte frequencies
      freq = Hash.new(0)
      data.each_byte do |b|
        freq[b] += 1
      end

      # Calculate entropy
      total = length.to_f
      entropy = 0.0
      
      freq.each do |_, count|
        p = count / total
        entropy -= (p * Math.log2(p)) if p > 0 && !p.nan?
      end

      @entropy = entropy.round(4)
      
      # Also calculate per-section entropy if PE headers found
      if @headers[:dos_found]
        sections_entropy = []
        
        @headers[:sections].each do |section|
          next unless section[:raw_offset] > 0 && 
                      section[:raw_offset] + section[:raw_size] <= length

          sec_data = data[section[:raw_offset], section[:raw_size]]
          
          if sec_data.length > 16
            # Calculate entropy for this section
            sec_freq = Hash.new(0)
            sec_data.each_byte do |b|
              sec_freq[b] += 1
            end

            sec_total = sec_data.length.to_f
            sec_entropy = 0.0
            
            sec_freq.each do |_, count|
              sp = count / sec_total
              sec_entropy -= (sp * Math.log2(sp)) if sp > 0 && !sp.nan?
            end

            sections_entropy << {
              :name => section[:name],
              :entropy => sec_entropy.round(4)
            }
          end

          self.entropy = sections_entropy
        end
      end

      self.entropy
    end

    # Generate YARA rules output
    def generate_yara!
      return "" if @headers.key?(:yara_output)

      data = @data.dup
      length = data.length
      
      yara_rules = []
      
      # UPX rule
      if @headers[:packers][:upx]
        upx_offsets = []
        [0, 64].each do |offset|
          next unless offset + 2 < length
          if [data[offset], data[offset+1]] == [76, 80] # "UP"
            upx_offsets << { :offset => offset }
          end
        end
        
        yara_rules << "rule UPX_Detector {"
        yara_rules << "    meta:"
        yara_rules << "        description = \"UPX packer detected\""
        yara_rules << "        author = \"Packpeek\""
        yara_rules << "        date = \"#{Time.now.strftime('%Y-%m-%d')}\""
        yara_rules << "        version = \"1.0\""
        yara_rules << "    condition:"
        
        if upx_offsets.any?
          offsets_str = upx_offsets.map { |o| o[:offset].to_s }.join(", ")
          yara_rules << "        (std::string(data, 0, 2) == \"UPX!\")"
          yara_rules << "        || (std::string(data, #{upx_offsets.first[:offset]}, 2) == \"UPX!\")"
        else
          yara_rules << "        std::string(data, 0, 2) == \"UPX!\""
        end
        
        yara_rules << "}"

        self.yara_output = yara_rules.join("\n") + "\n\n"
      end

      # ASPack rule
      if @headers[:packers][:aspack]
        aspack_offsets = []
        [0, 64].each do |offset|
          next unless offset + 7 < length
          if [data[offset], data[offset+1], data[offset+2], 
              data[offset+3], data[offset+4], data[offset+5], 
              data[offset+6]] == [83, 72, 73, 70, 76, 79, 87] # "ASPack"
            aspack_offsets